Feature stm32g4 can support

m-ecry opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->
Added CAN support for the STM32G4.
Added support of FDCAN3 interface for the STM32 family.
Added fdcan_core_clk calculation instead of basing the calculation on a fixed 10MHz.
<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->
None.
<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->
I've tested it with the NUCLEO-G474RE board & FDCAN-Adapter, don't know exactly how to do this section.
<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->
@jeromecoutant 
<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

m-ecry committed:
make G4 target compileable with CAN support

m-ecry committed:
Changed PLL to 160MHz, PLLQ to 80MHz

m-ecry committed:
Calculate FDCAN_clk instead of assuming fix 10MHz

 - The FDCAN_clk is calculated on runtime from the according
RCC-registers

m-ecry committed:
Added support for FDCAN3

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32G4/TARGET_STM32G474xx/TARGET_NUCLEO_G474RE/system_clock.c:
```diff
@@ -105,9 +105,9 @@ uint8_t SetSysClock_PLL_HSE(uint8_t bypass)
+    RCC_OscInitStruct.PLL.PLLN = 80;
     RCC_OscInitStruct.PLL.PLLP = RCC_PLLP_DIV2;
-    RCC_OscInitStruct.PLL.PLLQ = RCC_PLLQ_DIV2;
+    RCC_OscInitStruct.PLL.PLLQ = RCC_PLLQ_DIV4;
```
So you decrease SystemCoreClock from 170 MHz to 160,
and you get PLLQ frequency at 80 MHz

Could you explain me again the reason ?
Thx

m-ecry replied:
With a core frequency of 170MHz, many CAN-frequencies in my tests could not be setup correctly. 

For example I worked at a frequency of 250k. With a core frequency of 170MHz this gave me:

| Field | Calculated Value | Register Value |
|------|-------|-----|
| ntq1 	| 510 (0x1FE)	| 253 (0xfd) |
| ntq2 	| 169 (0xA9)	| 40  (0x28) |
| sjw 	| 169 (0xA9)	| 40  (0x28) |

So the register could not store the value, the interface could not be setup properly.

The current limitation is the result of the api not using the prescaler (always at 1). 

I propose to use the prescaler, calculated this way:
```C
    int ntq = HAL_RCCEx_GetPeriphCLKFreq(RCC_PERIPHCLK_FDCAN) / hz;

    int nominalPrescaler = 1;
    // !When the sample point should be lower than 50%, this must be changed to
    // !IS_FDCAN_NOMINAL_TSEG2(ntq/nominalPrescaler), since
    // NTSEG2 and SJW max values are lower. For now the sample point is fix @75%
    while (!IS_FDCAN_NOMINAL_TSEG1(ntq/nominalPrescaler)){
        nominalPrescaler ++;
        if (!IS_FDCAN_NOMINAL_PRESCALER(nominalPrescaler)){
            error("Could not determine good nominalPrescaler. Bad clock value\n");
        }
    }
    ntq = ntq/nominalPrescaler;
```
See commit. 

Still there can't be guaranteed, that the resulting timing is within tolerance, so this will still require manual adaptation when using a strange value.

Even with that a frequency of 170MHz caused many setups to fail, so I reduced it to 80MHz, which worked just fine for most (not all) of my test-configurations. Unfortunately I not record my tests properly, I can repeat and record them to show it better, if needed.

jeromecoutant replied:
Thx for details.
Maybe we could code like this:
````
#if DEVICE_CAN
PLLN = 80
#else
PLLN = 85
````
If you disable CAN, you keep the maximum capability ?

m-ecry replied:
That fine by me, I will add this with a comment.

m-ecry replied:
Added in 2a13fa199dcc9f2a8faef79b7377e12222830d23, added comment in 73493b909a442d97f1d208c14620ca8d55636223.

jeromecoutant commented on targets/TARGET_STM/can_api.c:
```diff
@@ -118,7 +179,8 @@ static void _can_init_freq_direct(can_t *obj, const can_pinmap_t *pinmap, int hz
     */
-    int ntq = 10000000 / hz;
+
+    int ntq = _can_get_core_frequency() / hz;
```
Maybe you could try HAL_RCCEx_GetPeriphCLKFreq with RCC_PERIPHCLK_FDCAN ?

jeromecoutant replied:
Mmm it seems that STM32H7 doesn't provide the same function... :-(

jeromecoutant replied:
Note you can try code like this:
````
#ifdef TARGET_STM32G4
int ntq = HAL_RCCEx_GetPeriphCLKFreq(RCC_PERIPHCLK_FDCAN)/ hz;
#else
// STM32H7 doesn't support yet HAL_RCCEx_GetPeriphCLKFreq for FDCAN
// Internal ST ticket 92465
int ntq = 10000000 / hz;
#endif

````

m-ecry replied:
This is a really nice catch, thank you! I added your proposal in. (35c9e7a)

jeromecoutant commented on targets/TARGET_STM/can_api.c:
```diff
@@ -491,6 +573,17 @@ void FDCAN2_IT1_IRQHandler(void)
     can_irq(CAN_2, 1);
 }
 
+void FDCAN3_IT0_IRQHandler(void)
```
#if defined(FDCAN3_BASE) is missing

m-ecry replied:
I didn't add this line, since it was also missing for the FDCAN2 Handlers, which otherwise are also guarded with the #if defined(GDCAN2_BASE). 

But I agree, it makes sense to add the guard for both the FDCAN2 and FDCAN3 Handlers.

m-ecry replied:
Added guards in 35c9e7a

deveritec-marec committed:
Use HAL function for FDCAN_CLK-calculation

 - Thanks to @jeromecoutant for showing the HAL funtion
 - Added #ifdef guard to FDCAN2/3 handler functions

m-ecry committed:
STM-can-api: Added usage of prescaler

 - This enables more frequencies, but without regard to the accuracy.
May still require manual clock setup, to remain in tolerance window

m-ecry committed:
STM-can-api: Support reading of remote_msg

 - Previously a received msg was fixed of data_type

m-ecry committed:
STMG4-sys-clk: If can PLLQ=160MHz, else 170MHz

 - with 170MHz as can-core-frequency, the accuracy for many baudrates is
too low. 160MHz is better for a broad range of frequencies

jeromecoutant commented on targets/TARGET_STM/can_api.c:
```diff
@@ -235,8 +275,34 @@ int can_frequency(can_t *obj, int f)
+#else
+    // STM32H7 doesn't support yet HAL_RCCEx_GetPeriphCLKFreq for FDCAN
+    // Internal ST ticket 92465
+    int ntq = 10000000 / hz;
```
Not hz in this function but f...

m-ecry replied:
Sorry for that slip-up, I have no H7 to test, but should have noticed it anyway. Thanks for testing!

m-ecry replied:
Changed in 73493b9

jeromecoutant reviewed (changes requested):
Very good! I could perform some CAN tests:
- between NUCLEO_H743ZI2 (FDCAN) and NUCLEO_L476RG (CAN) in master branch
- between NUCLEO_H743ZI2 (FDCAN) and NUCLEO_L476RG (CAN) with this pull request  (and a small patch, see my comment)
- between NUCLEO_H743ZI2 (FDCAN) and NUCLEO_G474RE (FDCAN) with this pull request

m-ecry committed:
STM-can-api: Fixed variable name for H7

 - can_frequency uses f instead of hz for can frequency
 - Also added comment to system_clock
